### PR TITLE
Added Swish and Mish activations

### PR DESCRIPTION
**Pull request by thebhatman**
I have added the Swish and Mish activation functions. This resolves https://github.com/opencv/opencv/issues/15693

###This Pull request changes
[Feature addition] New activation functions Swish and Mish.

```
force_builders=Custom
buildworker:Custom=linux-4
build_image:Custom=ubuntu-cuda:18.04

build_image:Custom Mac=openvino-2019r3.0
test_modules:Custom Mac=dnn,python2,python3,java
```

### CODE REVIEW

File: modules/dnn/src/layers/elementwise_layers.cpp [hunk 1]
```diff
@@ -613,6 +613,186 @@ struct TanHFunctor
...
+        inps.getUMatVector(inputs);
+        outs.getUMatVector(outputs);
+        String buildopt = oclGetTMacro(inputs[0]);
+
+        for (size_t i = 0; i < inputs.size(); i++)
+        {
+            UMat& src = inputs[i];
+            UMat& dst = outputs[i];
+
+            ocl::Kernel kernel("MishForward", ocl::dnn::activations_oclsrc, buildopt);
```
Comment (dkurt) [inline modules/dnn/src/layers/elementwise_layers.cpp]: Missed kernel implementation?

**Comment by dkurt** [comment]
Hi! Thanks for contribution.

Please check that all the backends are valid (OpenCL, CUDA, Intel's Inference Engine). Otherwise keep only default C++ implementation.

File: modules/dnn/src/cuda4dnn/kernels/activations.hpp [hunk 2]
```diff
@@ -18,6 +18,12 @@ namespace cv { namespace dnn { namespace cuda4dnn { namespace kernels {
     template <class T>
     void tanh(const csl::Stream& stream, csl::Span<T> output, csl::View<T> input);
 
+    template <class T>
+    void swish(const csl::Stream& stream, csl::Span<T> output, csl::View<T> input);
+
+    template <class T>
+    void mish(const csl::Stream& stream, csl::Span<T> output, csl::View<T> input);
```
[the remark this answers is not in the document]
Comment (thebhatman) [reply modules/dnn/src/cuda4dnn/kernels/activations.hpp]: I have added the kernel implementation. Can you please help me with adding tests?

[the remark this answers is not in the document]
Comment (YashasSamaga) [reply modules/dnn/src/cuda4dnn/kernels/activations.hpp]: @thebhatman The CUDA kernels are called by a CPU function which selects the correct vectorized kernel based on the alignment.

You can have a look at [sigmoid](https://github.com/opencv/opencv/blob/85e848a0f73b69d99ae2e068f4987e2fc09c16fd/modules/dnn/src/cuda/activations.cu#L281-L302) example. The `sigmoid()` calls `launch_vectorized_sigmoid` which actually launches the kernel.

1. `SwishOp` calls `swish<T>(stream, output, input)`
2. `swish()` checks the alignment of `output` and `input` and calls `launch_vectorized_swish`
3. `launch_vectorized_swish` selects the correct kernel and launches it

[the remark this answers is not in the document]
Comment (thebhatman) [reply modules/dnn/src/cuda4dnn/kernels/activations.hpp]: @YashasSamaga Thank you! I have added the function for launching the kernel.

File: modules/dnn/src/cuda/activations.cu [hunk 3]
```diff
@@ -62,6 +62,44 @@ namespace cv { namespace dnn { namespace cuda4dnn  { namespace kernels {
...
+            auto input_vPtr = vector_type::get_pointer(input.data());
+
+            for (auto i : grid_stride_range(output.size() / vector_type::size())) {
+                vector_type vec;
+                v_load(vec, input_vPtr[i]);
+                for (int j = 0; j < vector_type::size(); j++) {
+                    using device::tanh;
+                    using device::log;
+                    using device::exp;
+                    vec.data[j] = vec.data[j] * tanh(log(1.0f + exp(vec.data[j])));
```
Comment (YashasSamaga) [inline modules/dnn/src/cuda/activations.cu]: Shouldn't `1.0f` be `T(1.0)`? This will enable some compiler optimizations in half precision kernels.

Comment (thebhatman) [reply modules/dnn/src/cuda/activations.cu]: Yes. That makes sense. Thank you!

**Comment by asmorkalov** [comment]
cc @VadimLevin

**Comment by dkurt** [comment]
@thebhatman, Please add the tests for both layers.

**Comment by thebhatman** [comment]
Yeah I will add the tests. Where exactly are the tests for activations? Should I define sample neural networks with these activations ? Or is there a test file with all activation functions such  as sigmoid and tanh?

**Comment by dkurt** [comment]
@thebhatman, So due the activations are simple, you can test them without test data. So you can just create a test which applies all the backends and compares with reference data (which is computed inside the test). Take a look at [test_layers.cpp](https://github.com/opencv/opencv/blob/master/modules/dnn/test/test_layers.cpp) and [test_halide_layers.cpp](https://github.com/opencv/opencv/blob/master/modules/dnn/test/test_halide_layers.cpp). The only parameter you need to vary is target and backend.

**Comment by thebhatman** [comment]
I have added the tests in `test_halide_layers.cpp` and they are passing. It seems like all tests in `test_layers.cpp` use Caffe Models to test layers by reading data from a file from opencv_extra. Is there something else to be done?

File: modules/dnn/test/test_halide_layers.cpp [hunk 4]
```diff
@@ -583,7 +583,7 @@ TEST_P(NoParamActivation, Accuracy)
     testInPlaceActivation(lp, backendId, targetId);
 }
 INSTANTIATE_TEST_CASE_P(Layer_Test_Halide, NoParamActivation, Combine(
-/*type*/ Values("TanH", "Sigmoid", "AbsVal", "BNLL"),
+/*type*/ Values("TanH", "Sigmoid", "AbsVal", "BNLL", "Swish", "Mish"),
```
Comment (dkurt) [inline modules/dnn/test/test_halide_layers.cpp]: Good choice! 😄

**Comment by thebhatman** [comment]
Is this PR ready to be merged now?

**Comment by YashasSamaga** [comment]
To enable the CUDA backend, you have to tick the following CMake options:
- `WITH_CUDA`
- `WITH_CUDNN`
- `OPENCV_DNN_CUDA`

The CUDA build is failing on CI. Please have a look at https://pullrequest.opencv.org/buildbot/builders/precommit_custom_linux/builds/2912

The CI does not run tests on CUDA devices. Hence, you (or someone) has to run them on their PC to verify that the tests are passing. Let me know when you are done with the PR, I'll build and test the PR once.

**Comment by thebhatman** [comment]
The error is because of the log function ` using device::log;`.

**Comment by thebhatman** [comment]
Thanks @YashasSamaga . The CUDA builds are now passing. They were failing due to type mismatch errors. I resolved them by using `log1pexp` instead of using log and exp separately.

File: modules/dnn/test/test_halide_layers.cpp (hunk 4 above)
Comment (YashasSamaga) [reply modules/dnn/test/test_halide_layers.cpp]: The CUDA tests are passing for the Swish activation.

The test for `DNN_TARGET_CUDA` is passing for the Mish activation. The test for `DNN_TARGET_CUDA_FP16` is failing due.

```
[ RUN      ] Layer_Test_Halide/NoParamActivation.Accuracy/21, where GetParam() = ("Mish", CUDA/CUDA_FP16)
.../opencv/modules/dnn/test/test_common.impl.hpp:68: Failure
Expected: (normL1) <= (l1), actual: 0.0288897 vs 0.004
.../opencv/modules/dnn/test/test_common.impl.hpp:71: Failure
Expected: (normInf) <= (lInf), actual: 0.0909604 vs 0.02
```

You will have to increase the error tolerance for Mish activation when target is `DNN_TARGET_CUDA_FP16`. I can't think of an easy and correct way to do this. In the worst case, the test can be skipped for the FP16 target.

**Comment by dkurt** [comment]
@thebhatman, there is no still OpenCL kernels implementation. Please add it or remove unused `applyOCL`.

File: modules/dnn/test/test_halide_layers.cpp (hunk 4 above)
Comment (thebhatman) [reply modules/dnn/test/test_halide_layers.cpp]: The default Thresholds are being called from here https://github.com/thebhatman/opencv/blob/f6221044661dedbf4ad729c8f24c2b72ff37aca0/modules/dnn/test/test_common.hpp#L113. I don't see a way to pass the thresholds as arguments without changing the prototype of `testInPlaceActivation` .

Comment (YashasSamaga) [reply modules/dnn/test/test_halide_layers.cpp]: I think you should add new parameters (`l1` and `lInf`) to `test` and `testInPlaceActivation`. Their default values would be `0.0`.

Inside the `test` function:
- if `l1` and `lInf` arguments are zero, set them to the values given by `getDefaultThresholds`
- if `l1` and `lInf arguments are not zero, use them

@dkurt is this  ok?

Comment (dkurt) [reply modules/dnn/test/test_halide_layers.cpp]: @YashasSamaga, yes, that would be fine. Thanks!

File: modules/dnn/src/layers/elementwise_layers.cpp [hunk 5]
```diff
@@ -613,6 +613,186 @@ struct TanHFunctor
...
+    {
+        Halide::Var x("x"), y("y"), c("c"), n("n");
+        top(x, y, c, n) = input / (1.0f + exp(-input));
+    }
+#endif  // HAVE_HALIDE
+
+#ifdef HAVE_INF_ENGINE
+    InferenceEngine::Builder::Layer initInfEngineBuilderAPI()
+    {
+        return InferenceEngine::Builder::SwishLayer("");
```
Comment (dkurt) [inline modules/dnn/src/layers/elementwise_layers.cpp]: There is no `SwishLayer` in this namespace. Please remove IE backend support for now.

Comment (thebhatman) [reply modules/dnn/src/layers/elementwise_layers.cpp]: I have removed the InfEngine backend support now. Thank you!

Comment (dkurt) [reply modules/dnn/src/layers/elementwise_layers.cpp]: Thanks!

Sorry, I forgot to highlight that you need to implement it in the following way (see `BNLLFunctor`):

```cpp
#ifdef HAVE_INF_ENGINE
    InferenceEngine::Builder::Layer initInfEngineBuilderAPI()
    {
        CV_Error(Error::StsNotImplemented, "");
    }
#endif  // HAVE_INF_ENGINE
```

Comment (thebhatman) [reply modules/dnn/src/layers/elementwise_layers.cpp]: Done. Thanks!

File: modules/dnn/src/layers/elementwise_layers.cpp [hunk 6]
```diff
@@ -613,6 +613,186 @@ struct TanHFunctor
...
+struct SwishFunctor
+{
+    typedef SwishLayer Layer;
+
+    bool supportBackend(int backendId, int)
+    {
+        return backendId == DNN_BACKEND_OPENCV ||
+               backendId == DNN_BACKEND_CUDA ||
+               backendId == DNN_BACKEND_HALIDE ||
+               backendId == DNN_BACKEND_INFERENCE_ENGINE;
```
Comment (dkurt) [inline modules/dnn/src/layers/elementwise_layers.cpp]: Please remove IE backend from this check and similar one in Mish layer.

Comment (thebhatman) [reply modules/dnn/src/layers/elementwise_layers.cpp]: I have removed it  now.

**Comment by cansik** [comment]
Has this already been released? In 4.3.0, I still get the error that it is unsupported:

```
OpenCV(4.3.0) /Users/travis/build/bytedeco/javacpp-presets/opencv/cppbuild/macosx-x86_64/opencv-4.3.0/modules/dnn/src/darknet/darknet_io.cpp:821: error: (-212:Parsing error) Unsupported activation: mish in function 'ReadDarknetFromCfgStream'
```

**Comment by dkurt** [comment]
@cansik, If I'm not mistaken, this fix for TensorFlow importer. Please track https://github.com/opencv/opencv/issues/17148

**Comment by YashasSamaga** [comment]
@BlueNotesRobot It is already there on master.
